gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/RequestInterceptor.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.tinkerpop.gremlin.driver;
+
+import org.apache.tinkerpop.gremlin.driver.handler.RequestInterceptorHandler;
+
+import java.util.function.UnaryOperator;
+
+/**
+ * This function is called when the {@link RequestInterceptorHandler} encounters an object passing through it.
+ * Implementers will need to detect the type of object to determine if they will interact with it or not. Typically,
+ * objects will be a {@code FullHttpRequest} in the case of a websocket handshake or some form of
```

Comment:
To begin with, shall we only restrict this to exposing the HTTP Requests Object and only at the stage of first handshake initialization with the server (that would mean that instead of being a handler this callback will be called from the place where we construct the first HTTP request)? Suggesting because invoking a function for every frame when the actual interceptor is required only for modifying the initial set of headers will have unnecessary performance penalty.

Reply:
I was trying to generalize this functionality rather than pin it particularly to `FullHttpRequest` and the initial handshake. We could say that's not useful and that only the first HTTP request matters I guess. This feature was largely your idea so I'll modify it to work the way you'd envisioned.